### PR TITLE
Initial commit for fortios_cmdb

### PR DESCRIPTION
**Pull request by migumun**
##### SUMMARY
FortiOS and FortiGate products need to be supported by Ansible, also for cmdb operations. This module aims to support that.

The module let the user configure different endpoints of the Fortigate, by allowing to set configuration values.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
fortios_cmdb_set

##### ANSIBLE VERSION
```
ansible 2.5.0 (fortios_cmdb 7782b9106e) last updated 2017/12/05 19:46:41 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/magonzalez/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/magonzalez/ansible/lib/ansible
  executable location = /home/magonzalez/ansible/bin/ansible
  python version = 2.7.14 (default, Sep 23 2017, 22:06:14) [GCC 7.2.0]
```

### CODE REVIEW

File: lib/ansible/modules/network/fortios/fortios_cmdb_set.py [hunk 1]
```diff
@@ -0,0 +1,728 @@
...
+from ansible.module_utils.basic import AnsibleModule
+import logging
+
+try:
+    from fortiosapi import FortiOSAPI
+    HAS_FORTIOSAPI = True
+except ImportError:
+    HAS_FORTIOSAPI = False
+
+if HAS_FORTIOSAPI:
```
Comment (gundalow) [inline lib/ansible/modules/network/fortios/fortios_cmdb_set.py]: Please delete this section

File: lib/ansible/modules/network/fortios/fortios_cmdb_set.py [hunk 2]
```diff
@@ -0,0 +1,728 @@
...
+        "endpoint": {"required": True,
+                     "choices": AVAILABLE_ENDPOINTS,
+                     "type": "str"},
+        "mkey": {"required": False, "type": "str"},
+        "config_parameters": {"required": True, "type": "dict"},
+    }
+
+    module = AnsibleModule(argument_spec=fields,
+                           supports_check_mode=False)
+    is_error, has_changed, result = fortios_cmdb_set(module.params)
```
Comment (gundalow) [inline lib/ansible/modules/network/fortios/fortios_cmdb_set.py]: Please add
```
try:
    from fortiosapi import FortiOSAPI
    HAS_FORTIOSAPI = True
except ImportError:
    HAS_FORTIOSAPI = False
```
before `is_error, has_changed, result = fortios_cmdb_set(module.params)`

File: test/runner/requirements/sanity.txt [hunk 3]
```diff
@@ -1,4 +1,5 @@
 cryptography
+fortiosapi
```
Comment (gundalow) [inline test/runner/requirements/sanity.txt]: Please remove this line.

This is only for what's common across all modules, notice no boto3 which is used by all AWS modules

File: lib/ansible/modules/network/fortios/fortios_cmdb_set.py (hunk 1 above)
Comment (migumun) [reply lib/ansible/modules/network/fortios/fortios_cmdb_set.py]: Thanks gundalow, seems to work now. I'll submit a new commit shortly.

File: lib/ansible/modules/network/fortios/fortios_cmdb_set.py [hunk 4]
```diff
@@ -0,0 +1,728 @@
...
+module: fortios_cmdb_set
+short_description: Module to configure all aspects of fortios and fortigate.
+description:
+    - This module is able to configure a fortios or fortigate by \
+    allowing the user to set every configuration endpoint from a fortios \
+    or fortigate device. The module transforms the playbook into \
+    the needed REST API calls.
+
+version_added: "2.5"
+author: "Nicolas Thomas / thomnico / (nthomas at fortinet.com) and \
```
Comment (gundalow) [inline lib/ansible/modules/network/fortios/fortios_cmdb_set.py]: List of Names & GitHub IDs please
```
author:
  - Nicolas Thomas (@thomnico)
  - Miguel Angel Munoz (@mamunozgonzalez)
```

File: lib/ansible/modules/network/fortios/fortios_cmdb_set.py [hunk 5]
```diff
@@ -0,0 +1,728 @@
...
+
+if HAS_FORTIOSAPI:
+    fos = FortiOSAPI()
+else:
+    raise ImportError("fortiosapi module is required")
+
+formatter = logging.Formatter('%(asctime)s %(name)-12s '
+                              '%(levelname)-8s %(message)s')
+logger = logging.getLogger('fortiosapi')
+hdlr = logging.FileHandler('/var/tmp/ansible-fortiosconfig.log')
```
Comment (gundalow) [inline lib/ansible/modules/network/fortios/fortios_cmdb_set.py]: Will this have issues if two or more instances or ansible-playbook are running at the same time?

**Review by gundalow — CHANGES_REQUESTED** [review]
Implementation seems different to othe rmodules (fortios_config.py) which build on top of  https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/network/fortios/fortios.py

File: lib/ansible/modules/network/fortios/fortios_cmdb_set.py (hunk 5 above)
Comment (migumun) [reply lib/ansible/modules/network/fortios/fortios_cmdb_set.py]: Basically the messages from both instances will end up in the same file. This is intended to provide a log for the fortiosapi library which is called by this module. Normally it is not a problem to have messages mixed (messages are very simple) but if required I can add a random id/cookie to every message printed out.

Comment (sivel) [reply lib/ansible/modules/network/fortios/fortios_cmdb_set.py]: I'm concerned with this being the default behavior.  In general we should not be creating files on users machines that didn't ask for.  Maybe a flag to enable this, but it should be off by default.  I'm also questioning why this should be set to `DEBUG` also.  Maybe it should only be enabled with `ANSIBLE_DEBUG` is enabled.

Regardless, I think this should be changed or removed.

**Review by sivel — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/fortios/fortios_cmdb_set.py (hunk 5 above)
Comment (migumun) [reply lib/ansible/modules/network/fortios/fortios_cmdb_set.py]: Regarding previous fortios_config.py, we wanted to complement the functionality given by that module. To do that we prefer to rely on Fortinet's fortiosapi library which is intended to have more generic support for this and the rest of features that will come next. Fortiosapi has been designed with FortiGate's REST API in mind, to simplify user experience (endpoints in playbook are mapped to a rest call that the user is already familiar with). And given the large REST API of the FortiGate, the module can be easily maintained and updated.

Comment (thomnico) [reply lib/ansible/modules/network/fortios/fortios_cmdb_set.py]: Agree on the default behavior it is a remains of the debugging developing. Will change that. Is there a better practice in Ansible to adapt to python logging we could follow ??

Comment (migumun) [reply lib/ansible/modules/network/fortios/fortios_cmdb_set.py]: Agree about the log. I'm removing it now. Will submit changes shortly.

**Comment by gundalow** [comment]
Implementation seems different to other modules (fortios_config.py) which build on top of https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/network/fortios/fortios.py

This module seems to want to do lots of things, and just be a wrapper round the API

**Comment by thomnico** [comment]
@gundalow on the wrapper to API. The code (in fortiosapi) is doing much more than wrapping the crud part of the rest API. I design it to be compatible with config management system. This set call is taking care of checking and applying the configuration whether the API call is a POST or a PUT. It also abstracts the vdom or global calls we can do our Firewall. The resulting playbook is a subset of the change one will want in the cmdb (endpoint) with proper error handling thanks to the API call.

**Comment by sivel** [comment]
We have recently passed guidelines against modules that try to do "too much".  This module does appear to be on that boarder.

The following comments appears in those guidelines:

* Modules should have a concise and well defined functionality. Basically, follow the UNIX philosophy of doing one thing well.
* Modules should not require that a user know all the underlying options of an api/tool to be used. For instance, if what the legal values for a required module parameter cannot be documented that's a sign the module would be rejected.

The `config_parameters` and `endpoint` effectively indicate to us that this module falls within that category.  The `config_parameters` argument is arbitrary, and are not an explicitly defined argument with set options.  This is offloading knowledge of how to do things onto the user, instead of abstracting away those details into an easy to use well defined module.

**Comment by thomnico** [comment]
You mean something like this: https://docs.ansible.com/ansible/2.4/aci_rest_module.html should not exist right ?

**Comment by sivel** [comment]
> You mean something like this: https://docs.ansible.com/ansible/2.4/aci_rest_module.html should not exist right ?

That is correct, and based on our recent additions to the guidelines, we have plans to go back and address and likely remove modules violating these new rules.  I don't have more details on that as of yet.

**Comment by sivel** [comment]
We did also recently decline a pull request for a module called `fortios_api` (https://github.com/ansible/ansible/pull/30767) which seems to have attempted something similar.

**Comment by thomnico** [comment]
in our case let say I do the 200 modules (1 per end point) with a description of every parameter .. the knowledge of the thing you need to configure (fortigate) is the same than passing a subset of parameters to the endpoint/API. The code will end up being more complex in general, more error prone and provide no benefit to our users or Ansible in general.  The required knowledge of the fortigate configuration is absolutely identical. So I only see problems going that route...

Let's have an example our proposal versus 350+ modules :

 - name:   firewall policy
     fortios_cmdb_set:
      endpoint: "firewall policy"
      host: "{{ host }}"
      username: "{{ username }}"
      password: "{{ password }}"
      vdom:  "{{  vdom }}"
      config_parameters:
         policyid: "66"
         name: "ansible"
         json:
           policyid: "66"
           name: "ansible"
           action: "accept"
           srcintf: [ {"name": "port1"} ]
           dstintf: [{"name":"port2"} ]
           srcaddr: [{"name":"all"} ]
           dstaddr: [{"name":"all"}]
           schedule: "always"
           service:  [{"name":"HTTPS"}]
           logtraffic: "all"

If it was 1 module:
 - name:   firewall policy
     fortios_cmdb_firewall_policy:
      host: "{{ host }}"
      username: "{{ username }}"
      password: "{{ password }}"
      vdom:  "{{  vdom }}"
       policyid: "66"
       name: "ansible"
       json:
         policyid: "66"
         name: "ansible"
         action: "accept"
         srcintf: [ {"name": "port1"} ]
         dstintf: [{"name":"port2"} ]
         srcaddr: [{"name":"all"} ]
         dstaddr: [{"name":"all"}]
         schedule: "always"
         service:  [{"name":"HTTPS"}]
         logtraffic: "all"

The resulting 350+ splitted modules will not make the user any favor....

**Comment by thomnico** [comment]
fortios_api (#30767) was a simple passthrough forcing the playbook writer to "code" in the playbook crud type. The one we propose is coming from Fortinet and as explained earlier it is declarative made for config management.

**Comment by mattclay** [comment]
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/46509/3/tests

**Comment by migumun** [comment]
Hi again,
sorry for the late reply. Finally we have found some time to rework this. Last Friday I briefly talked with @gundalow for a first impression but I think it would be better to leave some thoughts here as well. 
The idea would be to create as many modules as this list: 

system, wireless-controller, firewall, webfilter, ips, web-proxy, wanopt, application, dlp spamfilter, log, vpn, certificate, user, dnsfilter, antivirus, report, waf, authentication, switch controller, endpoint-control and router

That list greatly simplifies the functions that the fortigate can do. Thus, we could create an ansible module for each one of them. 

To start off, I could propose an example for webfiltering function: fortios_webfilter. It could be something like this:

```yaml

EXAMPLES = '''
- hosts: localhost
  vars:
   host: "192.168.122.40"
   username: "admin"
   password: ""
   vdom: "root"
  tasks:
  - name: Configure url to be filtered by fortigate
    fortios_webfilter:
      host:  "{{  host }}"
      username: "{{  username}}"
      password: "{{ password }}"
      vdom:  "{{  vdom }}"
      webfilter_url:
        id: "5"
        url: "www.test45.com"
        type: "simple"
        action: "exempt"
        status: "enable"
        exempt: "pass"
        web-proxy-profile: ""
        referrrer-host: ""
        state: "present"
'''



EXAMPLES = '''
- hosts: localhost
  vars:
   host: "192.168.40.8"
   username: "admin"
   password: ""
   vdom: "root"
  tasks:
  - name: Configure content to be used with webfilter feature
    fortios_webfilter:
      host:  "{{  host }}"
      username: "{{  username}}"
      password: "{{ password }}"
      vdom:  "{{  vdom }}"
      webfilter_content:
        name: "online-casino"
        pattern-type: "wildcard"
        status: "enable"
        lang: "western"
        score: "150"
        action: "block"
        state: "present"
       
'''
```
This example would be for a webfilter module that implements a couple of functions: a) to set urls and b) to set content words, that can be blocked/allowed by the fortigate.

Do you think we can continue this approach?

Thanks,
@mamunozgonzalez

File: test/units/modules/network/fortios/test_fortios_cmdb_set.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import sys
+from nose.plugins.skip import SkipTest
```
Comment (maxamillion) [inline test/units/modules/network/fortios/test_fortios_cmdb_set.py:22]: We'll also need to switch to [pytest](https://docs.pytest.org/en/latest/) for iunit tests as documented [here](https://docs.ansible.com/ansible/devel/dev_guide/testing_units.html#structuring-unit-tests). The Ansible Testing Community Group is attempting to standardize tests moving forward. Thank you!

**Comment by NilashishC** [comment]
@mamunozgonzalez Could you please confirm if this PR is going any further or is good to close?

**Comment by thomnico** [comment]
@NilashishC it is Ansible community/review which rejects this proposal. It is good for us and actually used by our customers. On our end, it has been rejected for months.

**Comment by abenokraitis** [comment]
@thomnico @mamunozgonzalez Looks like there are still changes requested by @sivel and @gundalow , as well as a rebase for shippable to work.

**Comment by thomnico** [comment]
> @thomnico @mamunozgonzalez Looks like there are still changes requested by @sivel and @gundalow , as well as a rebase for shippable to work.

Please read my comment from Mar 13.

**Comment by thomnico** [comment]
> @mamunozgonzalez Could you please confirm if this PR is going any further or is good to close?

It has been rejected by you/Ansible committee, not us. Please properly reject it in github to reflect your decision

**Pull request closed without merging**